ukvm/ukvm_module_gdb.c

```diff
@@ -19,308 +19,223 @@
 
-    return ch;
+    for (i = 0; i < count; i++) {
+        ch = hex(*buf++) << 4;
+        ch = ch + hex(*buf++);
+        *mem++ = ch;
+    }
+    return mem;
 }
 
+static int wait_for_connect()
+{
+    int listen_socket_fd;
+    struct sockaddr_in server_addr, client_addr;
+    struct protoent *protoent;
+    struct in_addr ip_addr;
+    socklen_t len;
+    int opt;
 
-/************************************************************************/
-/* BUFMAX defines the maximum number of characters in inbound/outbound buffers*/
-/* at least NUMREGBYTES*2 are needed for register packets */
-#define BUFMAX (400 * 4)
-
-int remote_debug;
-/*  debug >  0 prints ill-formed commands in valid packets & checksum errors */
-
-static const char hexchars[] = "0123456789abcdef";
-
-/* Number of registers.  */
-#define NUMREGS        32
+    listen_socket_fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (listen_socket_fd == -1) {
+        err(1, "Could not create socket");
+        return -1;
+    }
+    
+    opt = 1;
+    if (setsockopt(listen_socket_fd, SOL_SOCKET, SO_REUSEADDR, &opt,
+                   sizeof(opt)) == -1)
+        err(1, "setsockopt(SO_REUSEADDR) failed");
+
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    server_addr.sin_port = htons(portno);
+     
+    if (bind(listen_socket_fd, (struct sockaddr *)&server_addr,
+             sizeof(server_addr)) == -1) {
+        err(1, "bind failed");
+        return -1;
+    }
+     
+    if (listen(listen_socket_fd , 0) == -1) {
+        err(1, "listen failed");
+        return -1;
+    }
+     
+    warnx("Waiting for a debugger. Connect to it like this:");
+    warnx("\tgdb --ex=\"target remote localhost:%d\" KERNEL", portno);
+     
+    len = sizeof(client_addr);
+    socket_fd = accept(listen_socket_fd, (struct sockaddr *)&client_addr, &len);
+    if (socket_fd == -1) {
+        err(1, "accept failed");
+        return -1;
+    }
 
-/* Number of bytes of registers.  */
-#define NUMREGBYTES (NUMREGS * 8)
+    close(listen_socket_fd);
 
-/* list is here: gdb/amd64-linux-nat.c */
-enum regnames {
-    RAX, RBX, RCX, RDX,
-    RSI, RDI, RBP, RSP,
-    R8, R9, R10, R11,
-    R12, R13, R14, R15,
-    RIP, EFLAGS, CS, SS,
-    DS, ES, FS, GS
-};
+    protoent = getprotobyname("tcp");
+    if (!protoent) {
+        err(1, "getprotobyname (\"tcp\") failed");
+        return -1;
+    }
 
-/*
- * these should not be static cuz they can be used outside this module
- */
-long registers[NUMREGS];
+    opt = 1;
+    if (setsockopt(socket_fd, protoent->p_proto, TCP_NODELAY, &opt,
+                   sizeof(opt)) == -1)
+        err(1, "setsockopt(TCP_NODELAY) failed");
 
-/***************************  ASSEMBLY CODE MACROS *************************/
-/*                                                                            */
+    ip_addr.s_addr = client_addr.sin_addr.s_addr;
+    warnx("Connection from debugger at %s\n", inet_ntoa(ip_addr));
```

Comment:
Left that one on purpose because the output felt a bit "crowded". Want me to remove it?